src/qref/verification.py

```diff
@@ -0,0 +1,145 @@
+
+    Args:
+        routine: Routine or program to be verified.
+    """
+    if isinstance(routine, SchemaV1):
+        routine = routine.program
+    problems = _verify_routine_topology(routine, is_root=True)
+    if problems:
+        return TopologyVerificationOutput(False, problems)
+    else:
+        return TopologyVerificationOutput(True, problems)
+
+
+def _verify_routine_topology(routine: RoutineV1, is_root: bool) -> list[str]:
+    problems = []
+    flat_graph = _get_flat_graph_from_routine(routine, path=None)
+    edge_list = []
+    for source, targets in flat_graph.items():
+        edge_list += [(source, target) for target in targets]
+
+    problems += _find_cycles(flat_graph)
+    problems += _find_disconnected_ports(routine, is_root)
+
+    for child in routine.children:
+        new_problems = _verify_routine_topology(child, is_root=False)
+        if new_problems:
+            problems += new_problems
+    return problems
+
+
+def _get_flat_graph_from_routine(routine, path) -> dict[str, list[str]]:
+    graph = defaultdict(list)
+    if path is None:
+        current_path = routine.name
+    else:
+        current_path = ".".join([path, routine.name])
+
+    input_ports = []
+    output_ports = []
+
+    for port in routine.ports:
+        if port.direction == "input":
+            input_ports.append(".".join([current_path, port.name]))
+        elif port.direction == "output":
+            output_ports.append(".".join([current_path, port.name]))
+
+    for connection in routine.connections:
+        source = ".".join([current_path, connection.source])
+        target = ".".join([current_path, connection.target])
+        graph[source].append(target)
+
+    for child in routine.children:
+        input_ports = []
+        output_ports = []
+
+        child_path = ".".join([current_path, child.name])
+        for port in child.ports:
+            if port.direction == "input":
+                input_ports.append(".".join([child_path, port.name]))
+            elif port.direction == "output":
+                output_ports.append(".".join([child_path, port.name]))
+
+        for input_port in input_ports:
+            graph[input_port].append(child_path)
+
+        graph[child_path] += output_ports
+
+    return graph
+
+
+def _find_cycles(edges) -> list[str]:
+    for node in list(edges.keys()):
+        visited: list[str] = []
+        problem = _dfs_iteration(edges, node, node, visited)
+        if problem:
+            return problem
+    return []
+
+
+def _dfs_iteration(edges, initial_node, node, visited):
+    if node != initial_node:
+        visited.append(node)
+    for neighbour in edges[node]:
+        if neighbour not in visited:
+            if neighbour == initial_node:
+                return [f"Cycle detected for node: {node}. Cycle: {visited}."]
+            problem = _dfs_iteration(edges, initial_node, neighbour, visited)
+            if problem:
+                return problem
```

Comment:
**issue**: I think the function here is an antipattern, as it has both a nontrivial return value AND a mutable input.

However, we can easily remedy this switching to iterative DFS (there are other possibilities too):

```python
def _dfs_iteration(adj_list, start_node) -> list[str]:
    to_visit = [start_node]
    visited = []
    predecessors = {}
    
    while to_visit:
        node = to_visit.pop()
        visited.append(node)
        for neighbour in adj_list[node]:
            predecessors[neighbour] = node
            if neighbour == start_node:
                # Reconstruct the cycle 
                cycle = [neighbour]
                while len(cycle) < 2 or cycle[0] != start_node:
                    cycle.append(predecessors[cycle[-1]]                    
                return [f"Cycle detected: {cycle[::-1]}"]
            if neighbour not in visited:                
                to_visit.append(neighbour)

Reply:
Indeed, the code is much cleaner now, thank you!